### ISSUE TITLE
cleaner shiny integration

### ISSUE THREAD

**Issue by hafen** [bokeh/rbokeh#25]
@timelyportfolio, could you look into making embedding into shiny more straightforward?  When running the scaffolding with htmlwidgets, I get the following boilerplate:

    #' Widget output function for use in Shiny
    #'
    #' @export
    rbokehOutput <- function(outputId, width = '100%', height = '400px'){
      shinyWidgetOutput(outputId, 'rbokeh', width, height, package = 'rbokeh')
    }
    
    #' Widget render function for use in Shiny
    #'
    #' @export
    renderRbokeh <- function(expr, env = parent.frame(), quoted = FALSE) {
      if (!quoted) { expr <- substitute(expr) } # force quoted
      shinyRenderWidget(expr, rbokehOutput, env, quoted = TRUE)
    }

It would be nice to have this part of the package so that there are dedicated functions for embedding.  Could you look into this?

**Comment by timelyportfolio** [comment]
Sure, forgot I said I would do that.

**Comment by ramnathv** [comment]
@hafen can you elaborate on what you mean by needing dedicated functions for embedding? At a very basic level, the default binding supports refresh of the json structure triggering a redraw. For two-way exchange of information, one would need to plug into the callbacks provided by `bokeh`. I would be happy to talk this through with you and @timelyportfolio, if it would be easier.

**Comment by hafen** [comment]
@ramnathv, I was referring to the fact that `rbokehOutput` and `renderRbokeh` somehow got left out of the package back from when I originally made the scaffolding, so I want to add them back in and make sure they work.  Probably as trivial as putting those functions in and documenting them, although there may need to be something special done because of the way I'm handling `print`.

But beyond this, I'm really excited about exploring more rich shiny interactions through bokeh callbacks as you mentioned, with ideas such as streaming updates to an existing plot, plot interactions, etc.

**Comment by ramnathv** [comment]
Ah. I get it now.

**Comment by timelyportfolio** [comment]
and the `as.tags` is a little different with `rbokeh` so I need to work through that also

**Comment by ramnathv** [comment]
Ideally, it would be good to have `rbokeh` work with the standard print mechanism so as to avoid issues with internal API changes in `htmlwidgets`. I will go over the code to see what necessitates the deviation from the standard printing mechanism provided by `htmlwidgets`.

**Comment by timelyportfolio** [comment]
`rbokeh:::plot.BokehFigure` and `rbokeh:::plot.BokehGridPlot` get it where it can be used with `htmltools`.  This [gist](https://gist.github.com/timelyportfolio/60c9676395f541d1d743) demonstrates that.  I think just need to define as.tags.bokeh .

**Comment by hafen** [comment]
Yeah I think it goes back to [this discussion](https://github.com/ramnathv/htmlwidgets/issues/59), with the idea of creating the htmlwidget object from the start instead of at the end when print is called.  I'm not sure how easy that would be, but worth a look.

**Comment by ramnathv** [comment]
I remember that discussion. I will take a look to assess how easy or difficult it is to accomplish that.

**Comment by smartinsightsfromdata** [comment]
@hafen @timelyportfolio  any news on making the library working with shiny?

**Comment by timelyportfolio** [comment]
looks like this https://github.com/bokeh/bokeh/issues/554 is causing trouble with htmlwidgets.  I'll keep working on it.

**Comment by timelyportfolio** [comment]
@hafen @ramnathv Since bokeh.js includes `jQuery`, I'm not sure this will ever work as expected.  I'm not sure if a jquery-less build of bokeh.js is possible.  https://github.com/bokeh/bokeh/issues/554

**Comment by bryevdv** [comment]
Are we using `.noconflict()` incorrectly? Or does it simply not solve this problem in the way I thought it would? Shiny integration is definitely important, so we will figure something out.

**Comment by jcheng5** [comment]
@bryevdv I haven't looked too deeply into it, but it looks like you're calling `.noconflict()` instead of `.noconflict(true)`, it might be worth a try with the latter if you haven't already...?

**Comment by bryevdv** [comment]
Yikes what a silly mistake if so. We will try that ASAP, although I am not sure how best to evaluate if it fixes the problem.

**Comment by jcheng5** [comment]
I'm sure between @timelyportfolio and I, we can figure out if it's fixed, if you can provide the bokeh.min.js.

**Comment by bryevdv** [comment]
I can make it tonight and get it to you guys tomorrow.

**Comment by bryevdv** [comment]
does this look reasonable?
```
# Make sure that we don't clobber any existing definition of $ (most
  # likely a previous version of jQuery.
  _oldJQ = window.$

  # binding the libs that bokeh uses so others can reference them
  Bokeh._                 = require("underscore")
  Bokeh.$                 = require("jquery").noConflict(true)
  Bokeh.Backbone          = require("backbone")

  # restore any previous definition of $
  window.$ = _oldJQ
```

**Comment by bryevdv** [comment]
FWIW I am seeing that `window.$` already has a value, identical to the loaded by `require` which makes me think this is still not correct/sufficient.

**Comment by bryevdv** [comment]
This seems to do the trick: http://requirejs.org/docs/jquery.html#noconflictmap will get you guys a version to play with tomorrow.

**Comment by bryevdv** [comment]
Here is a file to test with:

http://103bb120a2d7eb904e9c-fdd35fa42705c38b8a1ea1dc55ea880a.r30.cf1.rackcdn.com/bokeh-rtest.min.js

We are still working through some test failures related to some jq add on shims but I think it should work well enough for  what you need.

**Comment by jcheng5** [comment]
I'll take a look at this tomorrow, thanks!

**Comment by jcheng5** [comment]
Thanks for your help @bryevdv. I just took a look and I'm not seeing the results I'm expecting from the bokeh-rtest.min.js file.

Try this in an otherwise-empty HTML file:

```html
<script src="http://code.jquery.com/jquery-1.11.2.min.js"></script>
<script src="http://103bb120a2d7eb904e9c-fdd35fa42705c38b8a1ea1dc55ea880a.r30.cf1.rackcdn.com/bokeh-rtest.min.js"></script>
```

Open that file in Chrome, and in the JS console, evaluate `$` and `jQuery`. On the right hand of the output it'll show where the code comes from. In both cases you should see that it's coming from bokeh-rtest.min.js. However, if you then run `$.noConflict(true)` and try `$` and `jQuery` again, you'll see that they come from jquery-1.11.2.min.js.

On the other hand, if I just take a stock version of bokeh.min.js and `s/noconflict()/noconflict(true)/g` then I get the results I expect: both `$` and `jQuery` are coming from the standalone jQuery right off the bat, and the JS errors in the Shiny app go away. (The bokeh plot still doesn't actually appear--that's a separate problem though.)

**Comment by jcheng5** [comment]
@hafen @timelyportfolio: I was able to get a Shiny app to render, but only with the following fixes:

1. Replace `noconflict()` with `noconflict(true)` in bokeh.min.js, as mentioned in my previous comment
2. In my app I had to have my renderRbokeh expression return the result of rbokeh:::plot.BokehFigure; it was the only way I could figure out how to get an actual htmlwidget out of rbokeh. We can talk about the right way to factor this but I suspect you'll want to make a more involved renderRbokeh that knows how to take the different Bokeh classes and turn them to widget objects.
3. The first line of the `renderValue` in the JS should be `el.innerHTML = "";`, otherwise repeatedly rendering an rbokeh will cause each new plot to be appended below the previous ones, instead of replacing them.

Here's my test app: https://gist.github.com/jcheng5/92f3e0e63b88548684c4

**Comment by bryevdv** [comment]
There was another last minute fix, and it is available in a dev build:

http://bokeh.pydata.org/en/latest/docs/installation.html#developer-builds

Using this build I was able to see that `window.$` is undefined. Can you try again with this new version of BokehJS?

**Comment by hafen** [comment]
Thanks @jcheng5 and @bryevdv!  I'm battling a nasty bug right now and not firing on all cylinders, but I'll take a look soon.  @timelyportfolio, if you have the bandwidth and would like to go ahead and take a look at permanent solutions for Joe's fixes, that'd be great.  I suppose #2 brings up a question we've had before about a BokehFigure object also being an htmlwidget object from the start instead of when it's printed.

**Comment by timelyportfolio** [comment]
Thanks everyone for working so hard on this.  @jcheng5 & @ramnathv , is this as simple as defining `as.tags` for `BokehFigure` and `BokehGrid` that would just call https://github.com/bokeh/rbokeh/blob/master/R/print.R#L53-L68 and then use `as.tags.htmlwidget`?

**Comment by jcheng5** [comment]
@timelyportfolio `as.tags` wouldn't solve the problem for renderRbokeh, as we don't want to convert it to tags at that point--we want to grab the `x` value, dependencies, and some basic options, and send those to the client as JSON. So the widget object itself is what we want.

**Comment by timelyportfolio** [comment]
Yes, @jcheng5 I'm not sure what I was thinking.  So, is a custom [`shinyRenderWidget`](https://github.com/ramnathv/htmlwidgets/blob/master/R/htmlwidgets.R#L259-L284) the only way to avoid requiring a user to call `plot.BokehFigure` or `plot.BokehGridPlot` inside their shiny server?

**Comment by jcheng5** [comment]
Yes but that's fine, the whole purpose of having render* functions in Shiny was to facilitate this kind of conversion.

Even better would be to have figure() return an htmlwidget in the first place, but I understand if that's not something you guys want to bite off now.

**Comment by hafen** [comment]
So it sounds like making things htmlwidgets from the start will clean a lot of things up.  I'll start looking into this when I'm back to work.

**Comment by timelyportfolio** [comment]
Up to you.  The main issue was the `jQuery` conflict.  This is more easily handled even in current form.

**Comment by bryevdv** [comment]
Just checking in, did we establish that the new dev build version solves the jq issue?

**Comment by jcheng5** [comment]
I'm really sorry, I dropped the ball on that. I'm now trying to install bokeh and all the dependencies over crappy hotel wifi and/or 4G hotspot and not having much success unfortunately. Is this something you can just pull from your machine and send me? If you wouldn't mind, my email is joe@rstudio.com. Otherwise I'll try again when I am better connected.

**Comment by timelyportfolio** [comment]
I'll try it now.

**Comment by timelyportfolio** [comment]
Ok, it seems to work well with @jcheng5 test app and also still on my standalone tests/examples.  Let me get synced with master, and I'll submit a pull.

**Comment by bryevdv** [comment]
I think you guys are indicating that its working but just in case here is a CDN to the latest dev build:

http://cdn.pydata.org/bokeh/dev/bokeh-0.8.2dev.52491359.min.js

**Comment by timelyportfolio** [comment]
I pulled bokeh from master this morning, and did grunt to build css and js and included that in the pull.

**Comment by hafen** [comment]
@ramnathv, @jcheng5, @timelyportfolio: I've been looking into creating an htmlwidget at the time `figure()` is called.  It will be straightforward, except that in the end when the widget is printed, I still need a custom print method to make final preparations for the widget instance data.  There's no way around this as several things cannot be computed until it is known that the figure is ready to be viewed.  I am thinking that this requirement will mean that this approach will lead to the same problem.  Do you have any suggestions?

**Comment by ramnathv** [comment]
Can you provide an example of a computation that needs to happen prior to the plot being viewed? I am still trying to understand what makes this situation unique.

**Comment by ramnathv** [comment]
I am thinking, this can be made easy by making some tweaks to [toHTML](https://github.com/ramnathv/htmlwidgets/blob/master/R/htmlwidgets.R#L45-L99). In particular, the `widget_data` function can be allowed to be overruled by `rbokeh_data` so that the `rbokeh` package can do the pre-rendering computations necessary. This will work as long as all the necessary information is available in the payload `x`. @hafen an example of some computations that need to be carried out pre-rendering would be useful for me to think about this redesign.

**Comment by ramnathv** [comment]
Okay. I think I have some idea now, looking through your code. You have to  run the `prepare_*` functions prior to rendering. Let me see if there is a quick and easy fix to `htmlwidgets` to allow this.

**Comment by hafen** [comment]
Great!  Thanks.  The general idea is that I have an object I can potentially keep updating with new layers and it is easier to store / update that object in a way that is not the same format as needed for the final widget instance data.  So it's not until the end that I get it in the format it needs to be for BokehJS.  There are several small nuances, such as resolving color schemes (which you can't do until you know you have all the layers) and handling deferred layers (such as abline which needs to know the extents of the axes), etc.

**Comment by ramnathv** [comment]
@hafen, here is my question to you then. Is it possible to encapsulate all those calculations that need to be done pre-rendering into a single function, say `render_figure`. Then all that needs to be done is for `htmlwidgets` to allow this function to be run on `x` before passing it on to the `toHTML` function.

**Comment by hafen** [comment]
@ramnathv yes that will indeed be possible.

**Comment by ramnathv** [comment]
Great. So here is what I am thinking. I will try to refactor some functions in `htmlwidgets` that allows widget authors to specify a pre-render hook that modifies the object `x` before passing it on to the printing function. I would have to think through the exact API design, but one possibility is for `htmlwidgets` to look for a function named `rbokeh_data` (may have to come up with a better name) that will modify `x` before rendering. I will create a new branch on `htmlwidgets` to experiment with this and `rbokeh` can be a good testing ground.

**Comment by hafen** [comment]
Sounds perfect.  You could have an s3 method, something like `htmlwidget_prerender_hook` with the default method being the identity.  Then I can write whatever methods I like (even separate ones for BokehFigure vs. BokehGridPlot, etc.) knowing they will be executed if `x` has that class.

**Comment by ramnathv** [comment]
We considered S3 methods, but decided against it since it would require package authors to import the generic from `htmlwidgets`. We are using name based resolution as an alternative(eg. `rbokeh_html` if detected would override `widget_html`).

So, most likely, you would end up writing a `rbokeh_prerender` function that will do the pre rendering processing. You could write this function as an S3 method, but note that you would have to make sure that the appropriate class gets added.

**Comment by hafen** [comment]
Sounds great.  Can do that.

**Comment by pandurang-kolekar** [comment]
Great work!
Eagerly waiting for Shiny binding for rbokeh..

**Comment by smartinsightsfromdata** [comment]
@hafen Apologies but I'm a bit confused.  Is shiny working now with rbokeh - or a ```rbokeh_prerender``` as referred to by @ramnathv  needs to be developed in order to have shiny properly integrated?

**Comment by hafen** [comment]
@pandurang-kolekar, @smartinsightsfromdata, yes, proper integration is waiting on `rbokeh_prerender`, although it is possible although not very clean - @timelyportfolio, can you comment on this?

@ramnathv, is there an issue for the prerender functionality that I can track, or should I create one?  I'd offer to jump in to help add this feature to htmlwidgets but my bandwidth is a bit tight at the moment and you probably already know exactly what to do.

**Comment by ramnathv** [comment]
@hafen I have a local branch setup as of now and will be pushing it once it is is reasonable shape. I will cc you when I do that.

**Comment by hafen** [comment]
Thanks @ramnathv - I'm sure you're very busy as well!

**Comment by hafen** [comment]
@ramnathv, I've been looking at the code - could the hook simply be put at the beginning of `toHTML`?

**Comment by ramnathv** [comment]
@hafen I would rather not put things in `toHTML`. A clean solution would be to add a pre-render hook function to `createWidget`, that basically takes the widget payload as an argument and transforms it just prior to rendering. Then the `widget_data` function can be modified to run this hook before render. 

https://github.com/ramnathv/htmlwidgets/tree/feature/pre-render-hook

The basic idea is that in your `createWidget` call, you can specify a `preRenderHook` function as an argument. It will be run on the widget payload, prior to being rendered. For a minimal example, do the following

```r
devtools::create('newpkg')
setwd('newpkg')
htmlwidgets::scaffoldWidget('testwidget')
```

In the `createWidget` line of `testwidget`, add the following line

```R
createWidget(
  x,
  ...,
  preRenderHook = function(x){ paste("Hi", x$message) })
)
```

If you build the package now and run `testwidget('Hafen')`, it will render `Hi Hafen`. It works with Shiny as well.

I have pushed this branch to github, and tested it with a minimal example. We might have to iterate the design a few steps before settling on the final one. Let me know how it works for you.

**Comment by hafen** [comment]
Excellent!  Yes, that looks like the right way to do it.  I'll test it out!

**Comment by hafen** [comment]
@ramnathv, I wonder if you would consider the following change in

https://github.com/ramnathv/htmlwidgets/blob/feature/pre-render-hook/R/htmlwidgets.R#L126

changing:

```
x$x = x$preRenderHook(x$x)
```

to:
 
```
x = x$preRenderHook(x)
```

The reason for this:

In rbokeh, users build up a figure by adding layers and other components (legend, axes, etc.).  These are all available as functions that take the htmlwidget as input and return the modified htmlwidget.  This is nice for piping, etc.  In my preRenderHook function, I need to check to see if the user has left anything out for a given type of plot or something should be automatically added (for example, I automatically add a legend if they have any mapped aesthetic attributes), and then call the appropriate function (e.g. add_legend) to modify the object.  This means the hook function needs the entire htmlwidget object.  I have to defer these types of operations until plot time because I do not know until then which of these will be required and all the appropriate settings.

What do you think?

**Comment by ramnathv** [comment]
@hafen I don't see any reasons not to pass the entire widget object to `x`. For my widgets, I usually store everything within `x$x`, even when it is being added by pipe functions down the line (e.g. `x$x$xAxis` will contain the `xAxis` parameters), but I presume you are storing it directly on `x`, which necessitates this change.

I will run a few checks to make sure that this does not introduce any circular stuff, since `preRenderHook` sits inside `x` and also accepts it as an argument. It should not be a problem, but I just want to make sure.

So, do things work for you if I made this one change? Let me know.

**Comment by ramnathv** [comment]
I updated the branch with the change. Can you check if it serves your purpose? I might have to do a few more checks before sending a PR to the master branch, but I wanted to make sure that it serves the purpose it was initially designed for.

**Comment by hafen** [comment]
Thanks!  This is hopefully still natural for others.  I'll give it a try.

BTW, I actually am storing everything needed `x$x`, but I am calling some of my pipable functions inside of `preRenderHook`, and these functions, even though they only modify `x$x`, expect the entire htmlwidget as input and output.  Thus the only convenient way to call pipable functions in a hook is to get the whole object.

**Comment by ramnathv** [comment]
I don't think I understand. Can you provide a simple example to illustrate this?

**Comment by ramnathv** [comment]
Never mind. I think think I understand now. I had assumed that `preRenderHook`  was a separate function that did not use the piped-in functions, but it is a perfectly valid situation and I think this new design works with that.

**Comment by jcheng5** [comment]
This sounds perfect, we will need this too in leaflet (including the whole `x` as we may need to add HTML dependencies on the fly).

**Comment by hafen** [comment]
Just installed your feature branch and added the hook everything works like a charm!!  This is great - should clear up all the trouble with custom print methods (unnecessary now), shiny integration, etc.

**Comment by hafen** [comment]
Many thanks to everyone on this thread. I think we have proper shiny integration:

```
library(shiny)
library(rbokeh)

ui = shinyUI(fluidPage(
  rbokehOutput('a_bokeh_plot' )
))

server = function(input, output){
  output$a_bokeh_plot <- renderRbokeh({
    figure() %>% ly_points(1:10) 
  })
}

shinyApp(ui = ui, server = server)
```

**Comment by timelyportfolio** [comment]
awesome job everyone !

**Comment by timelyportfolio** [comment]
Just for testing, @jcheng5 test with new code probably better.

```r
library("shiny")
library("rbokeh")
library("htmlwidgets")

ui <- fluidPage(
    rbokehOutput("rbokeh")
)

server <- function(input, output, session) {
    output$rbokeh <- renderRbokeh({
        # Use invalidateLater() and jitter() to add some motion
        invalidateLater(1000, session)
        figure() %>%
            ly_points(jitter(cars$speed), jitter(cars$dist))
    })
}

shinyApp(ui, server)
```

**Comment by pandurang-kolekar** [comment]
Thanks, it works! :+1: 
Will experiment around.
